resourcemanager/util/shard_pool_map.go

```diff
@@ -0,0 +1,80 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"sync"
+
+	"github.com/pingcap/errors"
+)
+
+const shard = 8
+
+func hash(key string) int {
+	return int(key[0]) % shard
+}
+
+// ShardPoolMap is a map with shard
+type ShardPoolMap struct {
+	pools [shard]poolMap
+}
+
+// NewShardPoolMap creates a shard pool map
+func NewShardPoolMap() *ShardPoolMap {
+	var result ShardPoolMap
+	for i := 0; i < shard; i++ {
+		result.pools[i] = newPoolMap()
+	}
+	return &result
+}
+
+// Add adds a pool to the map
+func (s *ShardPoolMap) Add(key string, pool *PoolContainer) error {
+	return s.pools[hash(key)].Add(key, pool)
+}
+
+// Iter iterates the map
+func (s *ShardPoolMap) Iter(fn func(pool *PoolContainer)) {
+	for i := 0; i < shard; i++ {
+		s.pools[i].Iter(fn)
+	}
+}
+
+type poolMap struct {
+	mu      sync.Mutex
```

Comment:
Can we change this to sync.RWMutex? to make better read performance?

Reply:
Done